jujusvg.go

```diff
@@ -45,28 +46,52 @@ func NewFromBundle(b *charm.BundleData, iconURL func(*charm.Reference) string, f
 	}
 	sort.Strings(serviceNames)
 	services := make(map[string]*service)
+	servicesNeedingPlacement := make(map[string]bool)
 	for _, name := range serviceNames {
+		needsPlacement := false
 		serviceData := b.Services[name]
 		x, xerr := strconv.ParseFloat(serviceData.Annotations["gui-x"], 64)
 		y, yerr := strconv.ParseFloat(serviceData.Annotations["gui-y"], 64)
 		if xerr != nil || yerr != nil {
-			return nil, errgo.Newf("service %q does not have a valid position", name)
+			if serviceData.Annotations["gui-x"] == "" && serviceData.Annotations["gui-y"] == "" {
+				needsPlacement = true
+				x = 0
+				y = 0
+			} else {
+				return nil, errgo.Newf("service %q does not have a valid position", name)
+			}
 		}
-		charmId, err := charm.ParseReference(serviceData.Charm)
+		charmID, err := charm.ParseReference(serviceData.Charm)
 		if err != nil {
 			// cannot actually happen, as we've verified it.
 			return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
 		}
-		icon := iconMap[charmId.Path()]
+		icon := iconMap[charmID.Path()]
 		svc := &service{
 			name:      name,
-			charmPath: charmId.Path(),
+			charmPath: charmID.Path(),
 			point:     image.Point{int(x), int(y)},
-			iconUrl:   iconURL(charmId),
+			iconUrl:   iconURL(charmID),
 			iconSrc:   icon,
 		}
 		services[name] = svc
-		canvas.addService(svc)
+		if needsPlacement {
+			servicesNeedingPlacement[name] = true
```

Comment:
you could move this into the conditional above and then drop the `needsPlacement` entirely.